src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs

```diff
@@ -376,11 +382,29 @@ internal bool ShouldRestartWorkerChannel(string runtime, bool isWebHostChannel,
             return string.Equals(_workerRuntime, runtime, StringComparison.InvariantCultureIgnoreCase) && (isWebHostChannel || isJobHostChannel);
         }
 
-        private async Task RestartWorkerChannel(string runtime, string workerId)
+        private async Task RestartWorkerChannel(string runtime)
         {
+            if (_disposing || _disposed)
+            {
+                return;
+            }
             if (_languageWorkerErrors.Count < ErrorEventsThreshold)
             {
-                await InitializeJobhostLanguageWorkerChannelAsync(_languageWorkerErrors.Count);
+                try
+                {
+                    // Issue only one restart at a time.
+                    await _restartWorkerProcessSLock.WaitAsync();
+                    await InitializeJobhostLanguageWorkerChannelAsync(_languageWorkerErrors.Count);
+                }
+                finally
+                {
+                    // Wait before releasing the lock to give time for the process to startup and initialize.
+                    await Task.Delay(_restartWait);
```

Comment:
Changes look OK!  Instead of exposing disposed, it would help to pass cancellationToken to [Task.Delay](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.delay?view=netcore-3.1#System_Threading_Tasks_Task_Delay_System_TimeSpan_System_Threading_CancellationToken_)

Can you give it a try and see if setting the cancellation token in "Dispose" will do the trick?

Reply:
I think Task.Delay with a cancellationToken would throw an exception if task is canceled before timeout... How about TaskCompletionSource<bool> disposeTask that gets set (and therefore runs to completion) when dispose is invoked? Other conditions can check to see if this task ran to completion or not... and `await Task.Delay(_restartWait)` can become `await Task.WhenAny(Task.Delay(_restartWait), disposeTask)` followed by a check below to see if disposeTask's status is completed?

Reply:
Take a look at discussion [here ](https://stackoverflow.com/questions/20509158/taskcanceledexception-when-calling-task-delay-with-a-cancellationtoken-in-an-key) . Try the following:

```
Task.Delay(_restartWait).ContinueWith(tsk => {
   //code to run after the delay goes here
}, cancellationToken.Token);

```

Reply:
Anyway, that still seemed to throw an exception so reverted it.